### ISSUE TITLE
Opening a TextField causes stateful widgets to reinitialize

### ISSUE THREAD

**Issue by ThinkDigitalRepair** [flutter/flutter#31150]
There are many issues (closed) with this same topic including #26004, https://github.com/flutter/flutter/issues/20607
and questions on Stackoverflow. I finally got a stacktrace from the android side and I've only seen it once. Maybe this will help.
Also, it only happens on Android (Emulator or multiple physical devices), ~~not on iOS.~~
[Edit] it happens on iOS too.

 Video => https://streamable.com/ukjsb
```
> I/zygote  (27810): Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$OnUnhandledKeyEventListenerWrapper>: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/view/View$OnUnhandledKeyEventListener;
> I/zygote  (27810):   at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2203)
> I/zygote  (27810):   at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:637)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:518)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:466)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:140)
> I/zygote  (27810):   at void com.sangcomz.fishbun.ui.album.AlbumActivity.onCreate(android.os.Bundle) (AlbumActivity.java:70)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
> I/zygote  (27810):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
> I/zygote  (27810):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
> I/zygote  (27810):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
> I/zygote  (27810):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
> I/zygote  (27810):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
> I/zygote  (27810):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/zygote  (27810):   at void android.os.Looper.loop() (Looper.java:164)
> I/zygote  (27810):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
> I/zygote  (27810):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
> I/zygote  (27810):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
> I/zygote  (27810):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
> I/zygote  (27810): Caused by: java.lang.ClassNotFoundException: Didn't find class "android.view.View$OnUnhandledKeyEventListener" on path: DexPathList[[zip file "/data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/base.apk"],nativeLibraryDirectories=[/data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/lib/x86, /data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/base.apk!/lib/x86, /system/lib, /vendor/lib]]
> I/zygote  (27810):   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:125)
> I/zygote  (27810):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
> I/zygote  (27810):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
> I/zygote  (27810):   at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2203)
> I/zygote  (27810):   at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:637)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:518)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:466)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:140)
> I/zygote  (27810):   at void com.sangcomz.fishbun.ui.album.AlbumActivity.onCreate(android.os.Bundle) (AlbumActivity.java:70)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
> I/zygote  (27810):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
> I/zygote  (27810):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
> I/zygote  (27810):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
> I/zygote  (27810):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
> I/zygote  (27810):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
> I/zygote  (27810):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/zygote  (27810):   at void android.os.Looper.loop() (Looper.java:164)
> I/zygote  (27810):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
> I/zygote  (27810):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
> I/zygote  (27810):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
> I/zygote  (27810):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
> I/zygote  (27810): 
> I/zygote  (27810): Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$OnUnhandledKeyEventListenerWrapper>: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/view/View$OnUnhandledKeyEventListener;
> I/zygote  (27810):   at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2203)
> I/zygote  (27810):   at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:637)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:518)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:466)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:140)
> I/zygote  (27810):   at void com.sangcomz.fishbun.ui.album.AlbumActivity.onCreate(android.os.Bundle) (AlbumActivity.java:70)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
> I/zygote  (27810):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
> I/zygote  (27810):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
> I/zygote  (27810):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
> I/zygote  (27810):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
> I/zygote  (27810):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
> I/zygote  (27810):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/zygote  (27810):   at void android.os.Looper.loop() (Looper.java:164)
> I/zygote  (27810):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
> I/zygote  (27810):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
> I/zygote  (27810):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
> I/zygote  (27810):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
> I/zygote  (27810): Caused by: java.lang.ClassNotFoundException: Didn't find class "android.view.View$OnUnhandledKeyEventListener" on path: DexPathList[[zip file "/data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/base.apk"],nativeLibraryDirectories=[/data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/lib/x86, /data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/base.apk!/lib/x86, /system/lib, /vendor/lib]]
> I/zygote  (27810):   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:125)
> I/zygote  (27810):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
> I/zygote  (27810):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
> I/zygote  (27810):   at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2203)
> I/zygote  (27810):   at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:637)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:518)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:466)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:140)
> I/zygote  (27810):   at void com.sangcomz.fishbun.ui.album.AlbumActivity.onCreate(android.os.Bundle) (AlbumActivity.java:70)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
> I/zygote  (27810):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
> I/zygote  (27810):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
> I/zygote  (27810):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
> I/zygote  (27810):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
> I/zygote  (27810):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
> I/zygote  (27810):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/zygote  (27810):   at void android.os.Looper.loop() (Looper.java:164)
> I/zygote  (27810):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
> I/zygote  (27810):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
> I/zygote  (27810):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
> I/zygote  (27810):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
> I/zygote  (27810): 
> I/zygote  (27810): Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$OnUnhandledKeyEventListenerWrapper>: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/view/View$OnUnhandledKeyEventListener;
> I/zygote  (27810):   at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2203)
> I/zygote  (27810):   at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:637)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:518)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:466)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:140)
> I/zygote  (27810):   at void com.sangcomz.fishbun.ui.album.AlbumActivity.onCreate(android.os.Bundle) (AlbumActivity.java:70)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
> I/zygote  (27810):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
> I/zygote  (27810):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
> I/zygote  (27810):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
> I/zygote  (27810):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
> I/zygote  (27810):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
> I/zygote  (27810):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/zygote  (27810):   at void android.os.Looper.loop() (Looper.java:164)
> I/zygote  (27810):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
> I/zygote  (27810):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
> I/zygote  (27810):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
> I/zygote  (27810):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
> I/zygote  (27810): Caused by: java.lang.ClassNotFoundException: Didn't find class "android.view.View$OnUnhandledKeyEventListener" on path: DexPathList[[zip file "/data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/base.apk"],nativeLibraryDirectories=[/data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/lib/x86, /data/app/com.teddycare.flutterapp-Yf8m6oQV5Fwu45qHA1cevw==/base.apk!/lib/x86, /system/lib, /vendor/lib]]
> I/zygote  (27810):   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:125)
> I/zygote  (27810):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
> I/zygote  (27810):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
> I/zygote  (27810):   at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2203)
> I/zygote  (27810):   at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:637)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:518)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:466)
> I/zygote  (27810):   at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:140)
> I/zygote  (27810):   at void com.sangcomz.fishbun.ui.album.AlbumActivity.onCreate(android.os.Bundle) (AlbumActivity.java:70)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
> I/zygote  (27810):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
> I/zygote  (27810):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
> I/zygote  (27810):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
> I/zygote  (27810):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
> I/zygote  (27810):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
> I/zygote  (27810):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
> I/zygote  (27810):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/zygote  (27810):   at void android.os.Looper.loop() (Looper.java:164)
> I/zygote  (27810):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
> I/zygote  (27810):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
> I/zygote  (27810):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
> I/zygote  (27810):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
> I/zygote  (27810): 
> I/Choreographer(27810): Skipped 33 frames!  The application may be doing too much work on its main thread.
> E/RecyclerView(27810): No adapter attached; skipping layout
> D/EGL_emulation(27810): eglMakeCurrent: 0x9ec06c20: ver 3 0 (tinfo 0x8b151be0)
> W/Glide   (27810): Failed to find GeneratedAppGlideModule. You should include an annotationProcessor compile dependency on com.github.bumptech.glide:compiler in your application and a @GlideModule annotated AppGlideModule implementation or LibraryGlideModules will be silently ignored
```

**Comment by Hixie** [comment]
I don't understand the problem. Generally speaking, widgets should expect to rebuild all the time. What is the bug that this is causing?

**Comment by ThinkDigitalRepair** [comment]
Were you able to see the video?

**Comment by ThinkDigitalRepair** [comment]
I solved this by making the main widget use the AutomaticKeepAliveClientMixin. It initialized a TabController in its initstate and anytime the widget was destroyed, it would reset it to its default index. Closing this

**Comment by tvhieuit** [comment]
Here is my solution: 

`import 'package:flutter/material.dart';

class BaseMaterialPageRoute<T> extends MaterialPageRoute<T> {

  Widget result;

  BaseMaterialPageRoute({
    @required WidgetBuilder builder,
    RouteSettings settings,
    bool maintainState = true,
    bool fullscreenDialog = false,
  })  : assert(builder != null),
        assert(maintainState != null),
        assert(fullscreenDialog != null),
        assert(opaque),
        super(
            builder: builder,
            settings: settings,
            maintainState: maintainState,
            fullscreenDialog: fullscreenDialog);

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    if(result == null) {
      result = builder(context);
    }
    assert(() {
      if (result == null) {
        throw FlutterError(
            'The builder for route "${settings.name}" returned null.\n'
                'Route builders must never return null.'
        );
      }
      return true;
    }());
    return Semantics(
      scopesRoute: true,
      explicitChildNodes: true,
      child: result,
    );
  }
}
`

**Comment by SvenSchoene** [comment]
@tvhieuit This is absolutely crazy -- and it works! Thank you!

To clear his code up: the only thing he seemed to change was in the `buildPage`-method:

**Before:**
```dart
  @override
  Widget buildPage(
[...parameters...]
  ) {
    final Widget result = builder(context);
    assert(() {
      if (result == null) {
        throw FlutterError(
          'The builder for route "${settings.name}" returned null.\n'
          'Route builders must never return null.'
        );
      }
[...more code...]
  }
```

**After:** He took the `result` out of the method and made it a field of the class (and then initialized it in `buildPage()`, see:
```dart
  Widget result; // this is new

  @override
  Widget buildPage(
[...parameters...]
  ) {
    if (result == null) { // new
      result = builder(context); // new
    } // new
    assert(() {
      if (result == null) {
        throw FlutterError(
          'The builder for route "${settings.name}" returned null.\n'
          'Route builders must never return null.'
        );
      }
[...more code...]
  }
```

### Further information on the issue from my part

- The main factor behind the issue is using `onGenerateRoute` on the initial `MaterialApp` instead of setting up the `routes`-object.
- Going to any named route with a `TextFormField()` via `pushNamed()` then will cause any focus on the `TextFormField()` to cause the navigator to automatically go back one history state in the navigation stack. Doing some logging, this causes `build()` to be called on the previous Widget of the previous history state, as well as then navigating back to the page with the `TextFormField()`.

### Minimal reproduction of the issue

This is originally generated with `flutter create ...`, and then only adjusted as much as necessary to make the point:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      // home: MyHomePage(title: 'Flutter Demo Home Page'),
      // home: InitialPage(),
      onGenerateRoute: (RouteSettings settings) {
        switch (settings.name) {
          case "/":
            return MaterialPageRoute(
              builder: (BuildContext context) => InitialPage(),
              settings: settings,
            );
          case "/textfield-page":
            return MaterialPageRoute(
              builder: (BuildContext context) =>
                  MyHomePage(title: 'Flutter Demo Home Page'),
              settings: settings,
            );
        }
      },
    );
  }
}

class InitialPage extends StatelessWidget {
  Future<void> _redirect(BuildContext context) async {
    await Future.delayed(Duration(seconds: 2));
    Navigator.of(context).pushNamed("/textfield-page");
  }

  @override
  Widget build(BuildContext context) {
    _redirect(context);

    return Scaffold(
        body: Center(
      child: Text(
          "Initial Page. This is only here to navigate to another route after a few seconds."),
    ));
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextFormField(
              keyboardType: TextInputType.emailAddress,
              decoration: new InputDecoration(
                hintText: 'you@example.com',
                labelText: 'Email Address',
              ),
            ),
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

### Minimal change for a short-term fix

This is what @tvhieuit found out and implemented in the above example:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      // home: MyHomePage(title: 'Flutter Demo Home Page'),
      // home: InitialPage(),
      onGenerateRoute: (RouteSettings settings) {
        switch (settings.name) {
          case "/":
            return BaseMaterialPageRoute( // <-- NEW NEW NEW NEW NEW
              builder: (BuildContext context) => InitialPage(),
              settings: settings,
            );
          case "/textfield-page":
            return MaterialPageRoute(
              builder: (BuildContext context) =>
                  MyHomePage(title: 'Flutter Demo Home Page'),
              settings: settings,
            );
        }
      },
    );
  }
}

class InitialPage extends StatelessWidget {
  Future<void> _redirect(BuildContext context) async {
    await Future.delayed(Duration(seconds: 2));
    Navigator.of(context).pushNamed("/textfield-page");
  }

  @override
  Widget build(BuildContext context) {
    _redirect(context);

    return Scaffold(
        body: Center(
      child: Text(
          "Initial Page. This is only here to navigate to another route after a few seconds."),
    ));
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextFormField(
              keyboardType: TextInputType.emailAddress,
              decoration: new InputDecoration(
                hintText: 'you@example.com',
                labelText: 'Email Address',
              ),
            ),
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}

// NEW NEW NEW NEW NEW
// NEW NEW NEW NEW NEW
class BaseMaterialPageRoute<T> extends MaterialPageRoute<T> {
  Widget result;

  BaseMaterialPageRoute({
    @required WidgetBuilder builder,
    RouteSettings settings,
    bool maintainState = true,
    bool fullscreenDialog = false,
  })  : assert(builder != null),
        assert(maintainState != null),
        assert(fullscreenDialog != null),
        assert(opaque),
        super(
            builder: builder,
            settings: settings,
            maintainState: maintainState,
            fullscreenDialog: fullscreenDialog);

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    if (result == null) {
      result = builder(context);
    }
    assert(() {
      if (result == null) {
        throw FlutterError(
            'The builder for route "${settings.name}" returned null.\n'
            'Route builders must never return null.');
      }
      return true;
    }());
    return Semantics(
      scopesRoute: true,
      explicitChildNodes: true,
      child: result,
    );
  }
}

```

### Notes on this sample code

- I tested the code myself on a new project in Android emulator on Windows.
- `flutter doctor` result:
  ```
  Doctor summary (to see all details, run flutter doctor -v):
  [√] Flutter (Channel master, v1.5.9-pre.148, on Microsoft Windows [Version 10.0.17134.706], locale de-DE)
  [√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
  [!] Android Studio (version 3.1)
      X Flutter plugin not installed; this adds Flutter specific functionality.
      X Dart plugin not installed; this adds Dart specific functionality.
  [√] VS Code, 64-bit edition (version 1.33.1)
  [√] Connected device (1 available)

  ! Doctor found issues in 1 category.
  ```
- What did **not** help: Providing `TextFormField()` with a controller, which is part of the state of the stateful widget (meaning: `_MyHomePageState`).
- What did **not** help: Providing `TextFormField()` with a controller, which is part of the stateful widget (meaning: `MyHomePage`).
- What did **not** help: Using a `StatelessWidget` instead of a `StatefulWidget` for `MyHomePage`.
- What did **not** help: Providing a custom `focusNode` (focusNode either stored in `MyHomePage` or `_MyHomePageState`).
- What did **not** help: Wrapping everything in a `Form()`.
- What did **not** help: Wrapping everything in a `Form()` and providing a key for it, (key either stored in `MyHomePage` or `_MyHomePageState`).
- Please note that the new `BaseMaterialPageRoute` seems to only be necessary on the root-/previous navigation history state. I am not fully sure if it's either the root page that needs it, or the page that caused the `pushNamed()` (meaning: the previous history state).

I'm using the workaround that @tvhieuit found out for now -- however, I don't feel good about it because I don't know what this might break.  

I really hope this can either **be acknowledged as a bug** and therefore fixed as soon as possible (really meaning the "as possible", I don't mean "fix it instantly now!!1!!"), **or give an alternative solution and explanation where we all seem to be going wrong.**

In any case: thanks for the great work, Dart- and Flutter-Team! It's greatly appreciated, Flutter definitely is something special! :)

**Comment by SvenSchoene** [comment]
There is another crazy weird bug in the **fixed** example code: on every hot reload (I don't mean hot restart) the route/page is reloaded/rebuilt again. And it is stacking onto the navigation history. So, after x hot reloads the page with the form is x times on the navigation stack.

**Comment by SvenSchoene** [comment]
Ok, I feel extremely stupid now: the whole issue stems from the simple insight that when the keyboard is opened, the widget behind the current page is also being rebuilt (maybe even more levels behind it).

Since I had automatic redirection when the user is not logged in the widget behind the one with the input field, I was always being redirected back to the same navigation state.  

I guess everything works as intended then. Not sure how to prevent somebody else from going through this. But hey, maybe this also helps one other developer. :)

**Comment by SvenSchoene** [comment]
Ok, so I guess the reason why people have problems with this whole thing is because if one does not use named routes, opening the keyboard (and also hot reloading) does not call the builder of the previous page. See the full reproduction sample I'm adding to this comment.

### I guess this is the important question:
Is this a bug? Is this by design? What is the explanation for why the Widget-building behaves as it does when the keyboard is opened?
Clarifying this is probably going to bring some peace into this matter. :)

### Reproduction of what weirdly enough works:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: InitialPage(),
    );
  }
}

class InitialPage extends StatelessWidget {
  Future<void> _redirect(BuildContext context) async {
    await Future.delayed(Duration(seconds: 2));
    Navigator.push(
        context,
        MaterialPageRoute(
            builder: (context) => MyHomePage(title: 'Flutter Demo Home Page')));
  }

  @override
  Widget build(BuildContext context) {
    print("[InitialPage][build()]");
    _redirect(context);

    return Scaffold(
        body: Center(
      child: Text(
          "Initial Page. This is only here to navigate to another route after a few seconds."),
    ));
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextFormField(
              keyboardType: TextInputType.emailAddress,
              decoration: new InputDecoration(
                hintText: 'you@example.com',
                labelText: 'Email Address',
              ),
            ),
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

### And again, just to be absolutely clear: this does not work / produces the described erroneous behavior:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

String currentRoute = "";

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      onGenerateRoute: (RouteSettings settings) {
        currentRoute = settings.name;
        switch (settings.name) {
          case "/":
            return MaterialPageRoute(
              builder: (BuildContext context) => InitialPage(),
              settings: settings,
            );
          case "/textfield-page":
            return MaterialPageRoute(
              builder: (BuildContext context) =>
                  MyHomePage(title: 'Flutter Demo Home Page'),
              settings: settings,
            );
        }
      },
    );
  }
}

class InitialPage extends StatelessWidget {
  Future<void> _redirect(BuildContext context) async {
    await Future.delayed(Duration(seconds: 2));
    Navigator.of(context).pushNamed("/textfield-page");
  }

  @override
  Widget build(BuildContext context) {
    print("[InitialPage][build()]");
    _redirect(context);

    return Scaffold(
      body: Center(
        child: Text(
            "Initial Page. This is only here to navigate to another route after a few seconds."),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    print("[_MyHomePageState][build()]");

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: ListView(
          // mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextFormField(
              // controller: controller,
              keyboardType: TextInputType.emailAddress,
              decoration: new InputDecoration(
                hintText: 'you@example.com',
                labelText: 'Email Address',
              ),
            ),
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

**Comment by MaximCeban** [comment]
I'm facing the same issue, what is the solution for it?

**Comment by zbarryte-luminopia** [comment]
Though this is closed, this is still an issue.  I've opened a new issue, since this one is closed: https://github.com/flutter/flutter/issues/37878 

I also found that creating a subclass of `MaterialPageRoute` that caches the built widget does work as a patch, but I haven't been using this solution long enough to know if there are any adverse effects.

**Comment by PhyoeKo** [comment]
I'm facing the same issue, what is the solution for it?

TextInputField inside Stream widget tree, when I try to type something on textfield..
Sreambuilder is rebuilding. I think when softkeyboard is appear, the state full widget is rebuild.
Anyone please Help Me... :'(
